# Exploring Hacker News Posts

Hacker News is a popular site within the technology and startup communities. It gets hundreds of thousands of visitors per day, with several post types, including:

- Ask HN: users ask the Hacker News community specific questions
- Show HN: users post things to Hacker News to show off a project, product or something interesting

We're going to explore the data set provided by Hacker News to answer a couple of questions:

1. Do Ask HN or Show HN receive more comments on average?
2. Do posts created at a certain time receive more comments on average?

The first thing we need to do is import the data set as a readable list.

In [1]:
from csv import reader
opened_file = open('hacker_news.csv')
read_file = reader(opened_file)
hn = list(read_file)
hn_header = hn[:1]
hn = hn[1:]
print(hn[:5])

[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


## Step One: Separate the Data

Because we're only interested in the Ask HN and Show HN post types, we're going to create two new lists containing only those post types.

To do this, we need to:
1. Create empty lists
2. Find both the uppercase and lowercase versions of both post type
3. Move the post type information into the newly-created posts

In [2]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    title = title.lower()
    if title.startswith('ask hn'):
        ask_posts.append(row)
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))
    
   
    

1744
1162
17194


In [3]:
print('Ask Posts:', ask_posts[:5])
print('\n')
print('Show Posts:', show_posts[:5])
print('\n')

Ask Posts: [['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]


Show Posts: [['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46'], ['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https:/

## Part Two: Analyzing Ask HN vs Show HN

Now that we've created separate lists for both Ask HN and Show HN posts, we can determine which post type receives more comments on average.

To do this, we'll find the average number of posts for each comment type.

In [4]:
total_ask_comments = 0

for post in ask_posts:
    total_ask_comments += int(post[4])

avg_ask_comments = total_ask_comments / len(ask_posts)
print(avg_ask_comments)
    

    

14.038417431192661


In [5]:
total_show_comments = 0

for post in show_posts:
    total_show_comments += int(post[4])
    
avg_show_comments = total_show_comments / len(show_posts)
print(avg_show_comments)

10.31669535283993


*Summary of Comparison:*

As you can see above, Ask HN has more comments per post on average at 14.04 comments per post vs Show HN's 10.32 comments per post.

Because Ask HN posts receive more comments, we're going to focus on this post type for the rest of our analysis.

## Step Three: Investigating Posting Times

Our next step is to answer whether posts created at a certain time are more likely to attract comments. To do this, we're going to:

1. Calculate the number of posts created in each our of the day along with the number of comments received
2. Calculate the average number of commetns ask posts receive by hour created

In [6]:
import datetime as dt

In [7]:
result_list = []

for post in ask_posts:
    result_list.append([post[6], int(post[4])])
    

In [8]:
counts_by_hour = {}
comments_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for row in result_list:
    date = row[0]
    comment = row[1]
    time = dt.datetime.strptime(date, date_format).strftime('%H')
    if time in counts_by_hour:
        counts_by_hour[time] += 1
        comments_by_hour[time] += comment
    else:
        counts_by_hour[time] = 1
        comments_by_hour[time] = comment

## Step Four: Determining Best Time to Post

We now have two new dictionaries: one to count how many posts were created for each hour of the day and one with the corresponding number of comments created at each hour.

Now, we can move forward with creating a list of lists with the hours during which posts were created and the average number of comments those posts received.

In [9]:
avg_by_hour = []

for hr in comments_by_hour:
    avg_by_hour.append([hr, comments_by_hour[hr] / counts_by_hour[hr]])
    
print(avg_by_hour)

[['07', 7.852941176470588], ['16', 16.796296296296298], ['04', 7.170212765957447], ['02', 23.810344827586206], ['13', 14.741176470588234], ['21', 16.009174311926607], ['23', 7.985294117647059], ['12', 9.41095890410959], ['19', 10.8], ['20', 21.525], ['06', 9.022727272727273], ['03', 7.796296296296297], ['15', 38.5948275862069], ['09', 5.5777777777777775], ['05', 10.08695652173913], ['22', 6.746478873239437], ['17', 11.46], ['10', 13.440677966101696], ['14', 13.233644859813085], ['00', 8.127272727272727], ['01', 11.383333333333333], ['11', 11.051724137931034], ['18', 13.20183486238532], ['08', 10.25]]


## Part Five: Sorting the Results

Now that we have the list we need of the average number of comments for each post by hour, we can make it more readable to find our answer of which hour is best to post

In [10]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
    
print(swap_avg_by_hour)

[[7.852941176470588, '07'], [16.796296296296298, '16'], [7.170212765957447, '04'], [23.810344827586206, '02'], [14.741176470588234, '13'], [16.009174311926607, '21'], [7.985294117647059, '23'], [9.41095890410959, '12'], [10.8, '19'], [21.525, '20'], [9.022727272727273, '06'], [7.796296296296297, '03'], [38.5948275862069, '15'], [5.5777777777777775, '09'], [10.08695652173913, '05'], [6.746478873239437, '22'], [11.46, '17'], [13.440677966101696, '10'], [13.233644859813085, '14'], [8.127272727272727, '00'], [11.383333333333333, '01'], [11.051724137931034, '11'], [13.20183486238532, '18'], [10.25, '08']]


In [11]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
print(sorted_swap)

[[38.5948275862069, '15'], [23.810344827586206, '02'], [21.525, '20'], [16.796296296296298, '16'], [16.009174311926607, '21'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [13.20183486238532, '18'], [11.46, '17'], [11.383333333333333, '01'], [11.051724137931034, '11'], [10.8, '19'], [10.25, '08'], [10.08695652173913, '05'], [9.41095890410959, '12'], [9.022727272727273, '06'], [8.127272727272727, '00'], [7.985294117647059, '23'], [7.852941176470588, '07'], [7.796296296296297, '03'], [7.170212765957447, '04'], [6.746478873239437, '22'], [5.5777777777777775, '09']]


In [12]:
print('Top 5 Hours for Ask Posts Comments')

for avg, hr in sorted_swap[:5]:
    print(
        "{}: {:.2f} average comments per post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"),avg
        )
    )

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


# Conclusion

After narrowing down the top five average comments by post and tying it with the hour, we can conclude that 3pm EST/ 2pm CST is the best time to post if you want to have higher engagement.